protobuf-format.md

```diff
@@ -0,0 +1,387 @@
+contains the binary representation of the embedded message and a URL
+string that identifies the type of the embedded message.
+
+
+#### 2.2.2 Handling instructions
+
+If the `contentType` is `application/json` or any media type with the
+structured `+json` suffix, the implementation MUST store the JSON
+payload in the `data` field.
+
+If the payload is binary, the implementation MUST store the bytes in
+the `bytes_data` field.
+
+If the payload is a protobuf, the implementation MUST store the
+payload in the `proto_data` field and the `contentType` MUST be
+`application/protobuf`.
+
+### 2.3 Extensions:
+
+`google.protobuf.Struct` ([more info][PROTO_STRUCT]) represents an
+arbitrary JSON structure. The keys of the `Struct` are the names of
+the extensions. The values of the `google.protobuf.Struct` are
+`com.protobuf.Value` ([more info][PROTO_VALUE]) fields whose contents
+correspond to the value of the extension mapped into JSON using the
+[JSON mapping rules][CE_JSON_ENCODING]. All extensions attributes MUST
+be put into this bag because they are not a part of the CloudEvents
+spec.
+
+Well known extensions can not be given a top level field because they
+have no official standing, therefore the protobuf definition can not
+commit to a strongly typed representation of the value. The protobuf
+binding must be able to map an extension name to different value
+types, e.g. a "exampleExtension" with CE type "Object" in one message
+and another "exampleExtension" with CE type "Integer" in another
+message.
+
+When a new CloudEvent spec version is released, promoted extensions
+move from the extensions bag to a strongly typed top level field.
+
+For example, consider an extension for CloudEvents 1.0:
+
+```java
+CloudEvent.newBuilder()
+  .setCloudEventsVersion("1.0")
+  .setExtensions(
+    Struct.newBuilder()
+    .putFields(
+      "comExampleMyextension",
+      Value.newBuilder().setStringValue("myvalue").build()))
+  .build();
+```
+
+In the next CloudEvents minor release, the extension is promoted to an
+official field. The protobuf message definition is updated for
+CloudEvents 1.1:
+
+```java
+CloudEvent.newBuilder()
+  .setCloudEventsVersion("1.1")
+  .setComExampleMyextension("myvalue")
+  // During the upgrade process producers SHOULD continue setting
+  // exts bag to maintain backwards compat
+  .setExtensions(
+    Struct.newBuilder()
+      .putFields(
+        "comExampleMyextension",
+        Value.newBuilder().setStringValue("myvalue").build()))
+  .build();
+```
+
+#### 2.3.1 CloudEvents upgrade process
+
+If a field moves from the extensions bag to a top level field, then
+the producers and consumers of a CloudEvents system SHOULD coordinate
+in the upgrade process:
+
+1. Initially, the producers and consumers are using CloudEvents 1.0
+   and the extension is expressed in the “extensions” bag.
+1. CloudEvents 1.1 is released.
+1. The producers write the extension to both the extensions bag and
```

Comment:
Aside from the obvious concern about duplication of data, or worse, these two bits of data being different - can someone provide a URL to any spec out there that has a similar requirement? I can't think of any spec (JSON or otherwise) that requires this type of behavior and so I'd like to see how/why they do it. Since I don't think our spec is doing anything unusual I don't understand why our spec needs to have such special rules.